### PR TITLE
fix: `rpcv2Cbor` event stream requests sending incorrect `:content-type`

### PR DESCRIPTION
**Pull request by lauzadis**
## Description of changes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt [hunk 1]
```diff
@@ -34,7 +34,29 @@ class RpcV2CborTest {
...
+            input: PutFooStreamingInput
+        }
+        
+        structure PutFooStreamingInput {
+            room: String
+            messages: PublishEvents
+        }
+        
         // Model taken from https://smithy.io/2.0/spec/streaming.html#event-streams
+        @streaming
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:48]: Yes, PublishEvents / Message / LeaveEvent are in that Smithy docs page

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2Cbor.kt [hunk 2]
```diff
@@ -139,7 +139,12 @@ class RpcV2Cbor : AwsHttpBindingProtocolGenerator() {
         writer: KotlinWriter,
         resolver: HttpBindingResolver,
     ) {
-        writer.write("builder.headers.setMissing(\"Content-Type\", #S)", resolver.determineRequestContentType(op))
+        val contentTypeHeader = when {
+            op.isInputEventStream(ctx.model) -> "application/vnd.amazon.eventstream"
+            else -> "application/cbor"
+        }
+
+        writer.write("builder.headers.setMissing(\"Content-Type\", #S)", contentTypeHeader)
```
[the remark this answers is not in the document]
Comment (lauzadis) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2Cbor.kt:147]: The error we had before was in setting the `:content-type` during event stream message serialization. The value of that content type is set by `determineRequestContentType` which now only returns `application/cbor`. 

However the overall request `Content-Type` needs to be `application/vnd.amazon.eventstream` for event streams, which is what this `renderContentTypeHeader` function does

[the remark this answers is not in the document]
Comment (lauzadis) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2Cbor.kt:147]: I think the tests I wrote do a good job explaining the requirements but let me know if there's clarification to be done

File: codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt [hunk 3]
```diff
@@ -98,4 +120,29 @@ class RpcV2CborTest {
...
+        val encodeMessage = serializer.lines("private fun encodePutFooStreamingPublishEventsEventMessage(input: PublishEvents): Message = buildMessage {", "}")
+        encodeMessage.shouldContainOnlyOnceWithDiff(
+            """
+            addHeader(":content-type", HeaderValue.String("application/cbor"))
+            """.trimIndent(),
+        )
+
+        // Event stream requests should have Content-Type=application/vnd.amazon.eventstream
+        val serializeBody = serializer.lines("    override suspend fun serialize(context: ExecutionContext, input: PutFooStreamingRequest): HttpRequestBuilder {", "}")
+        serializeBody.shouldContainOnlyOnceWithDiff("""builder.headers.setMissing("Content-Type", "application/vnd.amazon.eventstream")""")
```
Comment (0marperez) [inline codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:146]: Question: This is coming from line 252 of `HttpBindingProtocolGenerator.kt` right?

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt [hunk 4]
```diff
@@ -249,6 +249,7 @@ abstract class HttpBindingProtocolGenerator : ProtocolGenerator {
                 if (op.isInputEventStream(ctx.model)) {
                     val eventStreamSerializeFn = eventStreamRequestHandler(ctx, op)
                     writer.write("builder.body = #T(context, input)", eventStreamSerializeFn)
+                    renderContentTypeHeader(ctx, op, writer, resolver)
```
Comment (0marperez) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt:252]: Question: This doesn't break anything for the other protocols?

File: codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (0marperez) [reply codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:48]: Nit: I think the comment is a little unnecessary

File: codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt (hunk 3 above)
Comment (lauzadis) [reply codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:146]: yes

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt (hunk 4 above)
Comment (lauzadis) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt:252]: not that I've seen. we have e2e tests for event streams in aws-sdk-kotlin

File: codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (lauzadis) [reply codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:48]: I disagree, more context is better than less

[the remark this answers is not in the document]
Comment (0marperez) [reply codegen/smithy-aws-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/RpcV2CborTest.kt:48]: That's fair

**Review by 0marperez — APPROVED** [review]

**Review by ianbotsf — APPROVED** [review]

**Pull request merged**